### ISSUE TITLE
Invalid environment variables whenever I used env function anywhere outside main server file

### ISSUE THREAD

**Issue by mudai11** [t3-oss/t3-env#134]
Trying to access `{ env }` anywhere outside the main fastify server file throws "Invalid environment variables" error so to solve this I have to use the library on the main file and send the env variables as arguments to functions that needs it for example `await injectPlugins(app, env.CORS_ORIGIN, env.SECRET_KEY);` I pass the fastify instance "app" and the two env variables so I can use them in another file to use cors origin for cors plugin and secret key for jwt plugin. Is there a way to use it in other files directly ?

Full code :
main.ts inside src folder :
```
import Fastify, { FastifyInstance } from "fastify";
import { Server } from "socket.io";
import { injectPlugins } from "./utils/plugins";
import { injectRoutes } from "./utils/routes";
import { injectSchemas } from "./utils/schemas";
import { env } from "../env";

export const app: FastifyInstance = Fastify();

declare module "fastify" {
  interface FastifyInstance {
    authenticate: any;
  }
}

declare module "fastify" {
  interface FastifyInstance {
    io: Server;
  }
}

async function build() {
  await injectPlugins(app, env.CORS_ORIGIN, env.SECRET_KEY);

  await injectRoutes(app);

  await injectSchemas(app);

  app.get("/", async function handler(_, reply) {
    reply.status(200).send("Welcome to my server! 🎉");
  });
}

async function main() {
  await build();

  try {
    app.listen({ port: 8080 }, (_, address) => {
      console.log(`server listening on: ${address}`);
    });
  } catch (e) {
    console.error("Server error: ", e);
    process.exit(1);
  }
}

main();
```

plugins.ts inside src/utils folder:

```
import { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";
import jwt from "@fastify/jwt";
import cookie from "@fastify/cookie";
import cors from "@fastify/cors";
import io from "fastify-socket.io";

export async function injectPlugins(
  app: FastifyInstance,
  cors_origin: string,
  secret_key: string
) {
  await app.register(cors, {
    origin: cors_origin,
    credentials: true,
  });

  await app.register(jwt, {
    secret: secret_key,
    cookie: {
      cookieName: "accessToken",
      signed: false,
    },
  });

  app.decorate(
    "authenticate",
    async (request: FastifyRequest, reply: FastifyReply) => {
      try {
        await request.jwtVerify();
      } catch (e) {
        reply.status(500).send(e);
      }
    }
  );

  await app.register(cookie, {
    secret: secret_key,
  });

  await app.register(io);
}
```

**Comment by mudai11** [comment]
> can you post a reproduction repo? might sound silly when you have provided code but if i should set up a project manually for each issue that quickly accumulates to time i don't have :/

here's the repo for the code : https://github.com/mudai11/chess-thing/tree/main/server

**Comment by juliusmarminge** [comment]
That repo seems to be private. Please provide a public one / Stackblitz etc

**Comment by mudai11** [comment]
@juliusmarminge sorry I made it private a few days ago, it's public again if you want to take a look at it.

https://github.com/mudai11/chess-thing/blob/main/server/src/main.ts

The problem lies in `injectPlugins()` inside `build()`, you can see that I'm passing env var as arguments because if I were to import `env` inside `injectPlugins()` that's inside `/utils/plugins.ts` file directly it won't work and gives me an "Invalid environment variables" error.

**Comment by deadcoder0904** [comment]
i faced the same error today. i was gonna open up an issue but glad i searched.

my error looks like:

> Unhandled Runtime Error
Error: Invalid environment variables

i was using `env.MODE` in `server` but wanted it on client as i wanted to set a client-cookie in browser in next.js so i moved it to `shared`.

```tsx
import { createEnv } from "@t3-oss/env-nextjs"
import { z } from "zod"

// see https://env.t3.gg/docs/nextjs
export const env = createEnv({
	server: {
		
	},
	client: {
		
	},
	shared: {
		MODE: z.enum(["development", "production", "staging"]),
	},
	experimental__runtimeEnv: {
		MODE: process.env.MODE,
	},
})
```

i include this component below in my global `layout.tsx`

```tsx
"use client"

import Cookie from "js-cookie"
import { useSearchParams } from "next/navigation"

import { env } from "@app/lib/env"

export function SetAffiliateCookie() {
	console.log(env.MODE)
	const searchParams = useSearchParams()
	const via = searchParams.get("via")
	if (via) {
		Cookie.set("hw_affiliate", via, {
			expires: 30,
			domain: env.MODE === "development" ? "localhost" : ".helloworld.com",
			secure: true,
			sameSite: "lax",
		})
	}

	return null
}
```

the funny part is it logs the above console perfectly as `development` but still gives runtime error.

```bash
 ❌ Invalid environment variables: { MODE: [ 'Required' ] }
 X node_modules/@t3-oss/env-core/dist/index.js (29:0) @ onValidationError
 X Error: Invalid environment variables
    at eval (./app/lib/env.ts:10:74)
    at (ssr)/./app/lib/env.ts (/home/projects/stackblitz-starters-agiu7i/.next/server/app/page.js:162:1)
    at __webpack_require__ (/home/projects/stackblitz-starters-agiu7i/.next/server/webpack-runtime.js:33:42)
    at eval (./app/components/Affiliate.tsx:8:66)
    at (ssr)/./app/components/Affiliate.tsx (/home/projects/stackblitz-starters-agiu7i/.next/server/app/page.js:151:1)
    at __webpack_require__ (/home/projects/stackblitz-starters-agiu7i/.next/server/webpack-runtime.js:33:42)
null
 ✓ Compiled in 1358ms
```

@juliusmarminge here's my repro -> https://stackblitz.com/edit/stackblitz-starters-agiu7i

**Comment by juliusmarminge** [comment]
Your issue is similar to this: https://github.com/t3-oss/t3-env/issues/209

tldr; shared envs should be used carefully and most of the times you're actually looking for a public client variable
